oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/query/ElasticRequestHandler.java

```diff
@@ -578,10 +588,72 @@ private boolean visitTerm(String propertyName, String text, String boost, boolea
     }
 
+    private ObjectBuilder<BoolQuery> inference(BoolQuery.Builder b, String propertyName, String text, PlanResult pr, boolean dbEnabled) {
+        ElasticIndexDefinition.InferenceDefinition.Query q = null;
+        // select first query eligible for the given text
+        // TODO: evaluate if/how to handle multiple queries
+        String  queryText = text;
+        for (ElasticIndexDefinition.InferenceDefinition.Query query : elasticIndexDefinition.inferenceDefinition.queries) {
+            if (query.isEligibleForInput(queryText)) {
+                queryText = query.rewrite(queryText);
+                if (query.hasMinTerms(queryText)) {
+                    q = query;
+                    break;
+                }
+            }
+        }
+
+        QueryStringQuery.Builder qsqBuilder = fullTextQuery(queryText, getElasticFieldName(propertyName), pr, dbEnabled);
+
+        // the query can be null if no inference query is eligible for the given text or the min terms are not met
+        // in this case, we fall back to the default full-text query
+        if (q != null) {
+            LOG.info("Using inference query: {}", q);
+            try {
+                // let's retrieve the fields with the same model as the query
+                final ElasticIndexDefinition.InferenceDefinition.Query query = q;
+                List<ElasticIndexDefinition.InferenceDefinition.Property> properties = elasticIndexDefinition.inferenceDefinition.properties.stream()
+                        .filter(pd -> pd.model.equals(query.model))
+                        .collect(Collectors.toList());
+                if (!properties.isEmpty()) {
+                    InferenceService inferenceService = InferenceServiceManager.getInstance(q.serviceUrl, q.model);
+                    List<Float> embeddings = inferenceService.embeddings(queryText, (int) q.timeout);
+                    if (embeddings != null) {
+                        for (ElasticIndexDefinition.InferenceDefinition.Property p : properties) {
+                            // https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-knn-query.html
+                            KnnQuery.Builder knnQueryBuilder = new KnnQuery.Builder();
+                            knnQueryBuilder.field(p.name + ".value");
+                            knnQueryBuilder.numCandidates(q.numCandidates);
+                            knnQueryBuilder.queryVector(embeddings);
+                            knnQueryBuilder.similarity(q.similarityThreshold);
+                            b.should(s -> s.knn(knnQueryBuilder.build()));
+                        }
+                        int tokens = text.split(" ").length;
+                        double qsBoost;
+                        if (tokens > 1) {
+                            qsBoost = 1.0d / (5 * tokens);
+                        } else {
+                            qsBoost = 1.0d;
+                        }
```

Comment:
It's a matter of taste, but here I think the ternary operator would be more clear.